### PR TITLE
swap and replace implementation

### PR DESCRIPTION
**Pull request by Drakulix**
@iwburns  input on the implementation please.
This time it is especially important, because there is a `TODO` left in the code.

### CODE REVIEW

File: src/tree.rs [hunk 1]
```diff
@@ -444,6 +444,49 @@ impl<T> Tree<T> {
...
+
+        let (is_valid, error) = self.is_valid_node_id(node_id);
+        if !is_valid {
+            return Result::Err(error.expect("Tree::replace: Missing an error value on finding an invalid NodeId."));
+        }
+
+        // lets cheat a bit and return new_node ;)
+        // TODO, while this is elegant, this worksaround NodeBuilder's Parameters
+        // Should we drain the old children array instead?
+        mem::swap(self.get_mut_unsafe(node_id).data_mut(), new_node.data_mut());
```
Comment (iwburns) [inline src/tree.rs]: So, this brings up an interesting question: why would people want to call replace?

Now that I think more about it, all we can really do is what you've done above because:
- we can't mess with the children/parent of the existing `Node` since it wouldn't really be a "replace" at that point.
- the `Node` that we return will need to have its `parent` and `child` cleared anyway to prevent `NodeId` leakage.

So it really all boils down to just swapping the data from one `Node` with the data from the other (unless I'm missing something here).

If that's the case we may want to reassess what this method is actually intended to be used for and possibly change it's name to reflect that a bit better (for example `replace_node_data(&mut self, id: &NodeId, data: T) -> T` might be more clear).

Thoughts?

**Comment by iwburns** [comment]
@Drakulix I commented in-line, I'm not sure if it pings you for that though, so I'm pinging you here.

File: src/tree.rs (hunk 1 above)
Comment (Drakulix) [inline src/tree.rs]: I just realized this as well. I think the actual use case is also already possible, because you may already do just `mem::swap(node.data_mut(), &mut some_new_data)`.

I am not sure anymore, if this library should actually provide `replace`.


My original idea was to move a leaf into the tree, which would work fine for existing nodes. But I quickly realized, that there is no good way to handle interior nodes this way, because it is not clear, what happens to their children in that case. So I moved to replacing with a new node, which defeats the original point somewhat.

What I wanted to achieve can already be done much better by using `swap` and then any of the `remove` functions.

So whats left of `replace` is a shortcut for `mem::swap`, which is not that useful anymore.

If you like `replace`, I would rather implement it on `Node` directly as in `Node::replace_data(&mut self, data: T) -> T`.

Comment (Drakulix) [inline src/tree.rs]: Well all `mem` functions are not exactly ergonomic.
I think its okay to provide a convenience function in that case.

File: src/tree.rs [hunk 2]
```diff
@@ -444,6 +444,49 @@ impl<T> Tree<T> {
...
+
+        Result::Ok(new_node)
+    }
+
+    pub fn swap(&mut self, first_id: &NodeId, second_id: &NodeId) -> Result<(), NodeIdError> {
```
Comment (iwburns) [inline src/tree.rs]: I'm having second thoughts about the name on this one.

To me (and maybe this is because of our discussion earlier about `mem::swap`) `swap` sounds like it would just swap the data in one `Node` with the data in another one (both of which are currently in the `Tree`).  While I think that would be a neat feature to have, that's not the intent with this function (correct?).

If we're moving the sub-tree around with the `Node` I almost want to call it `swap_with_children` or `swap_sub_trees` or something to that effect.  I'm not necessarily a fan of either of those, but I think they're more clear.  Thoughts?

Comment (Drakulix) [inline src/tree.rs]: I think `swap_sub_trees` is a good name. `swap_with_children` implies there is a swap without.

**Comment by Drakulix** [comment]
Rebased and fixed.
Pls review @iwburns

File: src/node.rs [hunk 3]
```diff
@@ -130,6 +130,29 @@ impl<T> Node<T> {
     }
 
     ///
+    /// Exchange the attached data through a new one
```
Comment (iwburns) [inline src/node.rs]: This is confusing wording to me.

How does this sound to you:
"Swaps the value of this `Node`'s data and the data provided."

I'm still not 100% happy with that wording either tbh, but what do you think?

File: src/node.rs [hunk 4]
```diff
@@ -130,6 +130,29 @@ impl<T> Node<T> {
...
+    ///
+    /// // ops! lets correct this
+    /// let three = node_four.replace_data(4);
+    ///
+    /// assert_eq!(node_four.data(), &4);
+    /// assert_eq!(three, 3);
+    /// ```
+    ///
+    pub fn replace_data(&mut self, mut data: T) -> T
+    {
```
Comment (iwburns) [inline src/node.rs]: I would prefer the `{` to be up one line with a single space preceding it.

File: src/node.rs (hunk 3 above)
Comment (Drakulix) [inline src/node.rs]: What about "Replaces this `Node`s data with the data provided".
The swap is explained in the next line and this also sounds wrong

File: src/node.rs (hunk 4 above)
Comment (Drakulix) [inline src/node.rs]: Sure thing,
you might be interested in adding a [rustfmt.toml](https://github.com/rust-lang-nursery/rustfmt) to your repo, because this might have happened more then one time to me and contributors could follow your guidelines more easily :)

File: src/node.rs [hunk 5]
```diff
@@ -173,6 +196,19 @@ impl<T> MutableNode for Node<T> {
         self.children.push(child);
     }
 
+    fn replace_child(&mut self, old: NodeId, new: NodeId) {
+        let mut index = 0;
+
+        for (i, id) in self.children().iter().enumerate()
+        {
+            if id == &old { index = i; break; }
+        }
```
Comment (iwburns) [inline src/node.rs]: I believe something like this would also work:
```
let index = self.children()
    .iter()
    .enumerate()
    .find(|(i, id)| id == &old).0;
```
(Those references might not be quite right, of course).

Would you be against that implementation instead of the manual index + break?

Comment (Drakulix) [inline src/node.rs]: I like the iterator patters very much, no idea, why I did not though about this.

Comment (iwburns) [inline src/node.rs]: No worries at all!

File: src/node.rs (hunk 4 above)
Comment (iwburns) [inline src/node.rs]: You're absolutely right, I've been meaning to do this and set up a CONTRIBUTING.md as well, but I just haven't done it yet.

**Comment by Drakulix** [comment]
All done

File: src/tree.rs [hunk 6]
```diff
@@ -436,6 +438,63 @@ impl<T> Tree<T> {
...
     ///
+    /// Swaps two `Node`s including their children given their `NodeId`s.
+    ///
+    /// Returns an empty `Result` containing a `NodeIdError` if one occurred on either provided `Id`.
+    ///
+    /// Notes:
+    ///     - Both `NodeId`s are still valid after this process and are not swapped.
+    ///     - Keeps the positions of the `Node`s in their parents children collection.
+    ///
+    /// ``
```
Comment (iwburns) [inline src/tree.rs]: I think you're missing a `` ` `` here.  I only noticed because I'm confused about your documentation test.

Maybe I'm missing something, but I think this won't work properly if one of the `Node`s being swapped is a direct descendant of the other `Node`.  In fact, I'm having a hard time wrapping my head around what _should_ happen in that scenario.

Also more generally, what is the expected behavior if one `Node` is in the sub-tree of the other `Node`?

Comment (Drakulix) [inline src/tree.rs]:
```
   A
 /   \
B     C
     /  \
    E    D
```
swap(A, C)
```
    C
  /   \
 B     A
     /   \
    E     D
```
^ Thats what I would expect and I actually have use cases requiring this behavior.

Comment (Drakulix) [inline src/tree.rs]: Also swapping children of the same node, should change their ordering.

Comment (iwburns) [inline src/tree.rs]: I assume you need the `NodeId` that points to `C` before the `swap` to still point to `C` after the swap correct (and the same would need to be true for `Node`-`A`)?
Because otherwise, I would think that you're essentially just swapping the `Node` data.

Comment (Drakulix) [inline src/tree.rs]: Ok now I got confused.
I though about the wrong function.

That actually a good question, I want to move the children, but I have no idea, what should happen, if one NodeId is in a sub tree of the other.
Thats most-likely the use case I had in mind when designing replace.
Move a child higher replacing a parent and dropping the rest in between.

Comment (Drakulix) [inline src/tree.rs]: swap(B, C):
```
         A
       /   \
      B     C
     / \   / \
    D   E F   G
```
```
         A
       /   \
      C     B
     / \   / \
    F   G D   E
```

swap(C, F):
```
         A
      /     \
    B         C
   / \       / \
  D   E     F   G
 / \ / \   / \ / \
H  I J  K L  M N  O
```
```
         A
      /     \
    B         F
   / \       / \
  D   E     L   M
 / \ / \
H  I J  K
```

This is what carefully documented I would like to see.
But I get that it is not exactly intuitive. Another option would be to add C into F's children minus the F sub tree, but that is more complicated to do.

The other way around (move upper `Node` into its own sub tree - swap(F, C) [order would become important here]) should error I my opinion. If we go with the alternative approach we could also lift the children of C and add C as child of F.

Comment (iwburns) [inline src/tree.rs]: Yeah, I think we should either prevent (somehow) the swapping of two `Node`s where one is in a sub-tree of the other, or we should put this off until we can come up with a reasonable behavior for that scenario (maybe looking at what other tree libraries do if they handle that kind of case?).

In addition though, I think it would be nice to have a version of `swap` that doesn't carry children with it to be honest.  It would be like swapping the data in two different `Node`s except that the `NodeId`s that refer to them would follow them to their new locations in the tree (meaning parent and child references would need to be updated of course).

Also, do you mind fixing the `` ` `` above?  I think it's preventing this doc test from running as I don't see it in here: 
image

Comment (Drakulix) [inline src/tree.rs]: Would you mind having a look at the behavior I suggested above?
I have fixed the ` locally. The result is a broken doctest...

I have added WIP label again and will report back, once those function are doing the things they should. But we need to decide how to handle the sub-tree-swapping.

Comment (Drakulix) [inline src/tree.rs]: Updated my graphics. Ping @iwburns

Comment (iwburns) [inline src/tree.rs]: I'm with you 100% on the `swap(B,C)` front.

As for the behavior described in `swap(C,F)` this would make more sense to me:
```
         A
      /     \
    B         C
   / \       / \
  D   E     F   G
 / \ / \   / \ / \
H  I J  K L  M N  O
```
would become:
```
         A
      /     \
    B         F
   / \       /|\
  D   E     C L M
 / \ / \     \
H  I J  K     G
             / \
            N   O
```
I'm thinking it should work that way because `F` was the first child of `C`, so `C` should now be the first child of `F`.  I also don't think order should matter here (calling `swap(C,F)` should be the same as calling `swap(F,C)`), but maybe I'm not thinking about that properly.  In addition, my first reaction is that `C`'s children shouldn't be dropped just because of a `swap`.

What do you think about that behavior?

Comment (Drakulix) [inline src/tree.rs]: Same as I outlined in my updated post :smile: 

It is fine, because if I want to emulate the behavior, I first suggested, I would just call `remove_node_drop_children(C)` (once that is merged) to get the desired effect.

Comment (iwburns) [inline src/tree.rs]: Alright, sounds good and no rush!

One last thing though, your example has `C` as the last child of `F` and mine has it as the first child of `F`.  On one hand I prefer mine because it's a very strict interpretation of `swap`, but on the other hand, your example doesn't disturb the places of `L` and `M` since it pushes `C` on the end of the array.

Not sure which way to go with this part because those are both desirable behaviors.

Comment (Drakulix) [inline src/tree.rs]: Well basically your reasoning is:  
> ... because F was the first child of C, so C should now be the first child of F.
While this is true and reasonable, I do not think, there is a use case for it.

Also what happens, if their is one layer between them? (Serious question, if you still want it to be implemented this way.)

I think just guaranteeing to add it to the end and not disturb the rest of the children is the best idea, because you want to keep the children, when swapping. Which means to me, that most likely you do not want to change the order of them by having a node inserted somewhere in the middle.

Comment (iwburns) [inline src/tree.rs]: I agree. Let's go with that.

I'm not really sure I'm wrapping my head around what the result would look like if we then swapped `F` and `G`.  Thoughts on that?  No rush.

Comment (Drakulix) [inline src/tree.rs]: I am really not looking forward to write tests for this... :confused:

Comment (iwburns) [inline src/tree.rs]: Yep, that makes sense to me. Thanks for drawing that up.

And yeah, writing tests for this won't be pretty.  If you want we can keep this off of Master for a bit too make sure we get the tests ironed out first. Up to you.

Let me know if you want me to set up a dev branch for this.

Comment (iwburns) [inline src/tree.rs]: Actually, a thought just came to mind: could we break a swap down into two `move_node_to_parent` calls (plus some extra logic and minor tweaks)?

Still on mobile so I can't draw it out but I think it might work fairly well. If it does, then we might be able to logic out the behavior a bit more easily.

Comment (Drakulix) [inline src/tree.rs]: This should indeed work for move _across_ the tree.
I am not sure how this would behave for sub tree moving. But that depends on solving #11 first anyway.

Comment (Drakulix) [inline src/tree.rs]: Alright, I did some testing.
Using `move_node_to_parent` two times actually works just nicely for moves *across*.

But for *down* and especially swapping the root break things spectacularly.
Lets assume the following example:
```
  A
 / \
B   C
|   |
D   E
```

`swap_sub_trees(D, E)` using `move_node_to_parent(E, D_Parent) ` and `move_node_to_parent(D, E_Parent) `
```
Step 1: (move_node_to_parent(E, D_Parent))
    A
   / \
  B   C
 / \
D   E

Step 2: (move_node_to_parent(D, E_Parent))
    A
   / \
  B   C
  |   |
  E   D
```

Swapping with the same parent also works with the correct order:
`swap_sub_trees(B, C)`
```
Step 1: (move_node_to_parent(C, B_Parent))
    A
   / \
  B   C
  |   |
  D   E

Step 2: (move_node_to_parent(B, C_Parent))
    A
   / \
  C   B
  |   |
  D   E
```

But breaks once we have more then two children.
This successfully swaps `B` and `C` but also moves `F` in the process:
`swap_sub_trees(B, C)`
```
     A
   / | \
  B  C  F
  |  |
  D  E

Step 1: (move_node_to_parent(C, B_Parent))
     A
   / | \
  B  F  C
  |     |
  D     E

Step 2: (move_node_to_parent(B, C_Parent))
     A
   / | \
  F  C  B
     |  |
     E  D
```

Lets get more interesting:
`swap_sub_trees(B, D)`
```
Step 1: (move_node_to_parent(D, B_Parent))
     A
   / | \
  B  C  D
     |
     E

Step 2: (move_node_to_parent(B, D_Parent))
This is the first time D_Parent is not entirely clearly defined.
If it is the parent at the start of the move this means move B under B (woot?).
If its the parent now we end up with:
     A
   / | \
  C  D  B
  |
  E

Expected:
    A
   / \
  C   D 
  |   |
  E   B
```
Not exactly what we wanted.

With the root involved this gets even more crazy:
`swap_sub_trees(A, D)`
```
Step 1: (move_node_to_parent(D, A_Parent))
A has no parent, does this mean we orphan D?
Lets just ignore this for now.

Step 2: (move_node_to_parent(A, D_Parent))
     B
     |
     D
     |
     A
     |
     C
     |
     E

Step 3: Maybe change the order and do  (move_node_to_parent(D, A_Parent)) now?
Still makes now sense. Either A_Parent is the original parent, which is still None,
or the Parent of A is A, so move_node_to_parent(A, A)??

Expected:
  D
 / \
B   C
|   |
A   E
```

I think this function needs a lot more logic then currently provided by `move_node_to_parent`, when we have a sub-tree match.
Even moves *across* the tree move is not perfect in terms of order (see the three children example), which leads me to believe even that should not be handled by any calls to `move_node_to_parent`.

If you have any more thoughts please tell me, I will most likely start with a clean new implementation now. My old one was to simple and your idea does not live up to our needs.

Comment (iwburns) [inline src/tree.rs]: On mobile, so I'm sorry for any typos.

I totally agree with you that we need a brand new (from scratch) solution to this, but I'm not really sure what it should be. 

My gut reaction is that it would be neat if we could just swap the data in the nodes and swap their children (the entire Vec), but then the NodeIds would be backwards right?  I'm not 100% sure, but wouldn't swapping the two Nodes in self.nodes fix that?

Comment (iwburns) [inline src/tree.rs]: Alright, I'm fairly certain that the above will work pretty well for _across_ moves.
Using `mem::swap` we could swap the following
- parent
- children
- data

After that, we need to remove those `NodeId`s from their original parents' children (and add them to their new parents' children), and then swapping the position of those two `Node`s in `self.nodes` should give us what we want. (I drew this out on paper and it seemed to make sense, but I haven't tried it in code yet).

What I'm not sure about is how to use similar logic to deal with up/down swaps.
In fact, now that I'm trying to think it through, I'm not even sure what we want it to do in the case where there's one or more `Node`s between the two being swapped (vertically).

Starting with:
```
   A
  / \
 B   C
    / \
   D   E
        \
         F
```
`swap(C,F)` could result in:
```
OPTION 1     OR     OPTION 2
   A                   A
  / \                 / \
 B   F               B   F
      \                   \
       E                   C
        \                 / \
         C               D   E
        /
       D
```
In `OPTION 1` above, we would keep the "swap parents" step but modify the "swap children" step (to leave `E` where it is instead of it traveling down with `C`).
In `OPTION 2` above, we would keep the "swap children" step but modify the "swap parents" step (to take 'E' down with 'C' which means that `E` then becomes the child of `F` instead of `E`).

As you can see this is very similar to the `move_node_to_parent` discussion.  Since we went with behavior similar to `OPTION 1` on that method, I'm thinking it makes sense to do the same here.  Do you have any thoughts on this @Drakulix ?

Comment (Drakulix) [inline src/tree.rs]: I dont think this makes sense here, and we already agreed on Option 2.

The reason it does not make more sense, is, that the swapped note should take their complete sub-tree with it, as long as that is possible to be as close to a real swap as we can.

You implementation idea sounds interesting, but I have already started another one. But I will keep that in mind.

Comment (iwburns) [inline src/tree.rs]: @Drakulix haha, ooops, I had forgotten that we agreed already on Option 2.  My bad!

And I agree actually, Option 2's behavior is probably what most people would expect here.

With those rules in mind, I just noticed that (unless I'm missing an edge case) if we're executing an up/down swap, the upper `Node` will always end up as the last child of the lower `Node`.  Is that right?

Comment (Drakulix) [inline src/tree.rs]: Yes that is right and makes that case most likely the easiest of them all, because you do not need to pay too much attention to the indices where you place the swapped notes.

Comment (iwburns) [inline src/tree.rs]: Man, that's a bit frustrating actually because `move_node_to_parent` is _almost_ that exact behavior.  Would be nice if we could just use that function, but it doesn't quite do what we want.

Comment (Drakulix) [inline src/tree.rs]: Right. Especially the up shift of move_node_to_parent gets in the way. Inserting at a different index then the end as well. we could probably split that logic even more apart, but I do not think we would gain much.

**Comment by Drakulix** [comment]
Alright, next try.

This is mostly an implementation of all the edge cases outline earlier.
It seems to pass all tests, but I am quite sure, this can be done more efficiently. I am happy however if this is just "correct" not efficient. We can iterate on this implementation later.

Please review @iwburns !

File: src/tree.rs [hunk 7]
```diff
@@ -625,6 +627,128 @@ impl<T> Tree<T> {
...
+            return Result::Err(error.expect("Tree::swap_sub_tree: Missing an error value on finding an invalid NodeId."));
+        }
+
+        let lower_upper_test = self.find_subtree_root_between_ids(first_id, second_id)
+        .map(|_| (first_id, second_id))
+        .or(
+            self.find_subtree_root_between_ids(second_id, first_id)
+            .map(|_| (second_id, first_id))
+        );
+
```
Comment (iwburns) [inline src/tree.rs]: Do you mind adding a hanging indent here?
I would prefer something like this:
```
let lower_upper_test = self.find_subtree_root_between_ids(first_id, second_id)
    .map(|_| (first_id, second_id))
    .or(
        self.find_subtree_root_between_ids(second_id, first_id)
        .map(|_| (second_id, first_id))
    );
```
I know this isn't documented anywhere, I promise I"ll get a rust-fmt style set up soon!  I just haven't had much time to get it set up lately.

File: src/tree.rs [hunk 8]
```diff
@@ -625,6 +627,128 @@ impl<T> Tree<T> {
...
+        if let Some((lower_id, upper_id)) = lower_upper_test
+        {
```
Comment (iwburns) [inline src/tree.rs]: Could you push `{` up to the previous line here?

File: src/tree.rs [hunk 9]
```diff
@@ -625,6 +627,128 @@ impl<T> Tree<T> {
...
+            let upper_parent_id = self.get_unsafe(upper_id).parent().cloned();
+
+            let lower_parent_id =
+            {
```
Comment (iwburns) [inline src/tree.rs]: I would like this `{` pushed up as well if you don't mind.

**Comment by Drakulix** [comment]
Done, done and done

File: src/tree.rs [hunk 10]
```diff
@@ -1147,4 +1270,34 @@ mod tree_tests {
         let sub_root = tree.find_subtree_root_between_ids(&root_id, &node_4_id); //invert for None
         assert_eq!(sub_root, None);
     }
+
+    #[test]
+    fn test_swap_sub_trees() {
+        let mut tree = Tree::new();
```
Comment (iwburns) [inline src/tree.rs]: I like these tests very much, but I feel like there is another case to test.  Can't think of anything off the top of my head though, so if you think this is exhaustive I'm fine with it.

Comment (Drakulix) [inline src/tree.rs]: The last case is probably in the example. But I can add some more. No problem.

Comment (iwburns) [inline src/tree.rs]: Any thoughts here?  @Drakulix no worries if you can't think of anything else to test. just trying to be thorough.

**Comment by Drakulix** [comment]
@iwburns Ping. More tests

**Comment by iwburns** [comment]
@Drakulix looks good to me, and feels more complete.

I'll merge and let you know when this is up in crates.io.

**Comment by Drakulix** [comment]
Wait some seconds with the push to crates.io please. I am updating the documation on the swap method a bit right now and I think this should be included.

**Comment by iwburns** [comment]
Ah, dang, it's up already.  Sorry!

**Comment by Drakulix** [comment]
Not too bad, lets continue here: #15
